File: domain-inventory/buggy/clojure/customers.clj
(ns customers)

(defrecord Address [street city state zip])

(defn address-street [r] (:street r))

(defn address-city [r] (:city r))

(defn address-state [r] (:state r))

(defn address-zip [r] (:zip r))

(defrecord Customer [id name email tier address created-year])

(defn customer-id [r] (:id r))

(defn customer-name [r] (:name r))

(defn customer-email [r] (:email r))

(defn customer-tier [r] (:tier r))

(defn customer-address [r] (:address r))

(defn customer-created-year [r] (:created-year r))

(defrecord PurchaseRecord [customer-id order-id amount year])

(defn purchaserecord-customer-id [r] (:customer-id r))

(defn purchaserecord-order-id [r] (:order-id r))

(defn purchaserecord-amount [r] (:amount r))

(defn purchaserecord-year [r] (:year r))

(def tier-gold "gold")

(def tier-silver "silver")

(def tier-bronze "bronze")

(defn tier-discount-pct [tier]
  (cond
  (= tier "gold") 5
  (= tier "silver") 10
  (= tier "bronze") 5
  true 0))

(defn tier-rank [tier]
  (cond
  (= tier "gold") 3
  (= tier "silver") 2
  (= tier "bronze") 1
  true 0))

(defn higher-tier? [t1 t2]
  (> (tier-rank t1) (tier-rank t2)))

(defn tier-from-spend [total-spend]
  (cond
  (>= total-spend 100000) "gold"
  (>= total-spend 50000) "silver"
  (>= total-spend 10000) "bronze"
  true "none"))

(defn find-customer-by-id [customers id]
  (first (filterv (fn [c] (= (customer-id c) id)) customers)))

(defn find-customer-by-email [customers email]
  (first (filterv (fn [c] (= (customer-email c) email)) customers)))

(defn customers-by-tier [customers tier]
  (filterv (fn [c] (= (customer-tier c) tier)) customers))

(defn customers-by-state [customers state]
  (filterv (fn [c] (= (address-state (customer-address c)) state)) customers))

(defn customers-by-city [customers city]
  (filterv (fn [c] (= (address-city (customer-address c)) city)) customers))

(defn sort-by-name [customers]
  (sort-by customer-name customers))

(defn sort-by-tier [customers]
  (sort-by (fn [c] (tier-rank (customer-tier c))) customers))

(defn sort-by-created [customers]
  (sort-by customer-created-year customers))

(defn customer-purchases [records cust-id]
  (filterv (fn [r] (= (purchaserecord-customer-id r) cust-id)) records))

(defn customer-total-spend [records cust-id]
  (reduce (fn [acc r] (+ acc (purchaserecord-amount r))) 0 (customer-purchases records cust-id)))

(defn customer-purchase-count [records cust-id]
  (count (customer-purchases records cust-id)))

(defn customer-avg-order-value [records cust-id]
  (let [purchases (customer-purchases records cust-id)
   cnt (count purchases)]
  (if (= cnt 0) 0 (quot (reduce (fn [acc r] (+ acc (purchaserecord-amount r))) 0 purchases) cnt))))

(defn customer-purchases-in-year [records cust-id year]
  (filterv (fn [r] (and (= (purchaserecord-customer-id r) cust-id) (= (purchaserecord-year r) year))) records))

(defn customer-spend-in-year [records cust-id year]
  (reduce (fn [acc r] (+ acc (purchaserecord-amount r))) 0 (customer-purchases-in-year records cust-id year)))

(defn assess-tier [records cust-id]
  (tier-from-spend (customer-total-spend records cust-id)))

(defn tier-upgrade-needed? [c records]
  (let [assessed (assess-tier records (customer-id c))
   current (customer-tier c)]
  (higher-tier? assessed current)))

(defn customers-needing-upgrade [customers records]
  (filterv (fn [c] (tier-upgrade-needed? c records)) customers))

(defn upgrade-customer-tier [c new-tier]
  (->Customer (customer-id c) (customer-name c) (customer-email c) new-tier (customer-address c) (customer-created-year c)))

(defn count-by-tier [customers]
  (let [gold (count (customers-by-tier customers "gold"))
   silver (count (customers-by-tier customers "silver"))
   bronze (count (customers-by-tier customers "bronze"))]
  {:gold gold :silver silver :bronze bronze}))

(defn total-customer-count [customers]
  (count customers))

(defn customers-since-year [customers year]
  (filterv (fn [c] (>= (customer-created-year c) year)) customers))

(defn customer-tenure [c current-year]
  (- current-year (customer-created-year c)))

(defn avg-customer-tenure [customers current-year]
  (let [cnt (count customers)]
  (if (= cnt 0) 0 (quot (reduce (fn [acc c] (+ acc (customer-tenure c current-year))) 0 customers) cnt))))

(defn top-spenders [customers records n]
  (vec (take n (reverse (sort-by (fn [c] (customer-total-spend records (customer-id c))) customers)))))

(defn most-active-customers [customers records n]
  (vec (take n (reverse (sort-by (fn [c] (customer-purchase-count records (customer-id c))) customers)))))

(defn customer-summary [c]
  (str (customer-name c) " (" (customer-tier c) ") - " (customer-email c)))

(defn customer-detail [c]
  (let [addr (customer-address c)]
  (str (customer-name c) " | " (customer-email c) " | Tier: " (customer-tier c) " | " (address-city addr) ", " (address-state addr))))

(defn address-oneline [a]
  (str (address-street a) ", " (address-city a) ", " (address-state a) " " (address-zip a)))

(defn valid-customer? [c]
  (and (> (customer-id c) 0) (not= (customer-name c) "") (not= (customer-email c) "")))

(defn valid-address? [a]
  (and (not= (address-street a) "") (not= (address-city a) "") (not= (address-state a) "") (not= (address-zip a) "")))

(defn unique-states [customers]
  (distinct (mapv (fn [c] (address-state (customer-address c))) customers)))

(defn state-customer-count [customers state]
  (count (customers-by-state customers state)))

(defrecord LoyaltyBalance [customer-id points lifetime-points])

(defn loyaltybalance-customer-id [r] (:customer-id r))

(defn loyaltybalance-points [r] (:points r))

(defn loyaltybalance-lifetime-points [r] (:lifetime-points r))

(defn points-for-purchase [amount tier]
  (let [base-points (quot amount 100)
   multiplier (cond
  (= tier "gold") 3
  (= tier "silver") 2
  (= tier "bronze") 1
  true 1)]
  (* base-points multiplier)))

(defn add-points [bal amount tier]
  (let [earned (points-for-purchase amount tier)]
  (->LoyaltyBalance (loyaltybalance-customer-id bal) (+ (loyaltybalance-points bal) earned) (+ (loyaltybalance-lifetime-points bal) earned))))

(defn redeem-points [bal points]
  (let [available (loyaltybalance-points bal)
   to-redeem (min points available)]
  (->LoyaltyBalance (loyaltybalance-customer-id bal) (- available to-redeem) (loyaltybalance-lifetime-points bal))))

(defn points-to-dollars [points]
  (quot points 10))

(defn find-loyalty-balance [balances cust-id]
  (first (filterv (fn [b] (= (loyaltybalance-customer-id b) cust-id)) balances)))

(defn top-loyalty-customers [balances n]
  (vec (take n (reverse (sort-by loyaltybalance-points balances)))))

(defn customer-segment [records cust-id]
  (let [spend (customer-total-spend records cust-id)
   count (customer-purchase-count records cust-id)]
  (cond
  (and (>= spend 100000) (>= count 10)) "vip"
  (and (>= spend 50000) (>= count 5)) "regular"
  (>= count 1) "occasional"
  true "inactive")))

(defn customers-by-segment [customers records segment]
  (filterv (fn [c] (= (customer-segment records (customer-id c)) segment)) customers))

(defn segment-distribution [customers records]
  {:vip (count (customers-by-segment customers records "vip")) :regular (count (customers-by-segment customers records "regular")) :occasional (count (customers-by-segment customers records "occasional")) :inactive (count (customers-by-segment customers records "inactive"))})

(defn years-since-last-purchase [records cust-id current-year]
  (let [purchases (customer-purchases records cust-id)]
  (if (empty? purchases) 999 (let [latest (reduce (fn [best r] (if (> (purchaserecord-year r) (purchaserecord-year best)) r best)) (first purchases) (rest purchases))]
  (- current-year (purchaserecord-year latest))))))

(defn at-risk-customers [customers records current-year threshold-years]
  (filterv (fn [c] (>= (years-since-last-purchase records (customer-id c) current-year) threshold-years)) customers))

(defn customer-retention-rate [customers records current-year threshold]
  (let [total (count customers)
   at-risk (count (at-risk-customers customers records current-year threshold))]
  (if (= total 0) 100 (quot (* (- total at-risk) 100) total))))

(defn update-customer-email [c new-email]
  (->Customer (customer-id c) (customer-name c) new-email (customer-tier c) (customer-address c) (customer-created-year c)))

(defn update-customer-address [c new-addr]
  (->Customer (customer-id c) (customer-name c) (customer-email c) (customer-tier c) new-addr (customer-created-year c)))

File: e8-scaled/trials/e8-python-run1/customers.py
from __future__ import annotations
from dataclasses import dataclass
import functools
from typing import Optional


@dataclass
class Address:
    street: str
    city: str
    state: str
    zip: str


@dataclass
class Customer:
    id: int
    name: str
    email: str
    tier: str
    address: Address
    created_year: int


@dataclass
class PurchaseRecord:
    customer_id: int
    order_id: int
    amount: int
    year: int


@dataclass
class LoyaltyBalance:
    customer_id: int
    points: int
    lifetime_points: int


TIER_GOLD = "gold"
TIER_SILVER = "silver"
TIER_BRONZE = "bronze"


def customer_id(r: Customer) -> int:
    return r.id


def customer_name(r: Customer) -> str:
    return r.name


def customer_email(r: Customer) -> str:
    return r.email


def customer_tier(r: Customer) -> str:
    return r.tier


def customer_address(r: Customer) -> Address:
    return r.address


def customer_created_year(r: Customer) -> int:
    return r.created_year


def tier_discount_pct(tier: str) -> int:
    if tier == "gold":
        return 15
    elif tier == "silver":
        return 10
    elif tier == "bronze":
        return 5
    else:
        return 0


def tier_rank(tier: str) -> int:
    if tier == "gold":
        return 3
    elif tier == "silver":
        return 2
    elif tier == "bronze":
        return 1
    else:
        return 0


def higher_tier(t1: str, t2: str) -> bool:
    return tier_rank(t1) > tier_rank(t2)


def tier_from_spend(total_spend: int) -> str:
    if total_spend >= 100000:
        return "gold"
    elif total_spend >= 50000:
        return "silver"
    elif total_spend >= 10000:
        return "bronze"
    else:
        return "none"


def find_customer_by_id(customers: list[Customer], id: int) -> Optional[Customer]:
    matches = [c for c in customers if c.id == id]
    return matches[0] if matches else None


def find_customer_by_email(customers: list[Customer], email: str) -> Optional[Customer]:
    matches = [c for c in customers if c.email == email]
    return matches[0] if matches else None


def customers_by_tier(customers: list[Customer], tier: str) -> list[Customer]:
    return [c for c in customers if c.tier == tier]


def customers_by_state(customers: list[Customer], state: str) -> list[Customer]:
    return [c for c in customers if c.address.state == state]


def customers_by_city(customers: list[Customer], city: str) -> list[Customer]:
    return [c for c in customers if c.address.city == city]


def sort_by_name(customers: list[Customer]) -> list[Customer]:
    return sorted(customers, key=lambda c: c.name)


def sort_by_tier(customers: list[Customer]) -> list[Customer]:
    return sorted(customers, key=lambda c: tier_rank(c.tier))


def sort_by_created(customers: list[Customer]) -> list[Customer]:
    return sorted(customers, key=lambda c: c.created_year)


def customer_purchases(records: list[PurchaseRecord], cust_id: int) -> list[PurchaseRecord]:
    return [r for r in records if r.customer_id == cust_id]


def customer_total_spend(records: list[PurchaseRecord], cust_id: int) -> int:
    return functools.reduce(lambda acc, r: acc + r.amount, customer_purchases(records, cust_id), 0)


def customer_purchase_count(records: list[PurchaseRecord], cust_id: int) -> int:
    return len(customer_purchases(records, cust_id))


def customer_avg_order_value(records: list[PurchaseRecord], cust_id: int) -> int:
    purchases = customer_purchases(records, cust_id)
    cnt = len(purchases)
    if cnt == 0:
        return 0
    return functools.reduce(lambda acc, r: acc + r.amount, purchases, 0) // cnt


def customer_purchases_in_year(records: list[PurchaseRecord], cust_id: int, year: int) -> list[PurchaseRecord]:
    return [r for r in records if r.customer_id == cust_id and r.year == year]


def customer_spend_in_year(records: list[PurchaseRecord], cust_id: int, year: int) -> int:
    return functools.reduce(lambda acc, r: acc + r.amount, customer_purchases_in_year(records, cust_id, year), 0)


def assess_tier(records: list[PurchaseRecord], cust_id: int) -> str:
    return tier_from_spend(customer_total_spend(records, cust_id))


def tier_upgrade_needed(c: Customer, records: list[PurchaseRecord]) -> bool:
    assessed = assess_tier(records, c.id)
    current = c.tier
    return higher_tier(assessed, current)


def customers_needing_upgrade(customers: list[Customer], records: list[PurchaseRecord]) -> list[Customer]:
    return [c for c in customers if tier_upgrade_needed(c, records)]


def upgrade_customer_tier(c: Customer, new_tier: str) -> Customer:
    return Customer(c.id, c.name, c.email, new_tier, c.address, c.created_year)


def count_by_tier(customers: list[Customer]) -> dict:
    gold = len(customers_by_tier(customers, "gold"))
    silver = len(customers_by_tier(customers, "silver"))
    bronze = len(customers_by_tier(customers, "bronze"))
    return {"gold": gold, "silver": silver, "bronze": bronze}


def total_customer_count(customers: list[Customer]) -> int:
    return len(customers)


def customers_since_year(customers: list[Customer], year: int) -> list[Customer]:
    return [c for c in customers if c.created_year >= year]


def customer_tenure(c: Customer, current_year: int) -> int:
    return current_year - c.created_year


def avg_customer_tenure(customers: list[Customer], current_year: int) -> int:
    cnt = len(customers)
    if cnt == 0:
        return 0
    return functools.reduce(lambda acc, c: acc + customer_tenure(c, current_year), customers, 0) // cnt


def top_spenders(customers: list[Customer], records: list[PurchaseRecord], n: int) -> list[Customer]:
    return list(sorted(customers, key=lambda c: customer_total_spend(records, c.id), reverse=True))[:n]


def most_active_customers(customers: list[Customer], records: list[PurchaseRecord], n: int) -> list[Customer]:
    return list(sorted(customers, key=lambda c: customer_purchase_count(records, c.id), reverse=True))[:n]


def customer_summary(c: Customer) -> str:
    return f"{c.name} ({c.tier}) - {c.email}"


def customer_detail(c: Customer) -> str:
    addr = c.address
    return f"{c.name} | {c.email} | Tier: {c.tier} | {addr.city}, {addr.state}"


def address_oneline(a: Address) -> str:
    return f"{a.street}, {a.city}, {a.state} {a.zip}"


def valid_customer(c: Customer) -> bool:
    return c.id > 0 and c.name != "" and c.email != ""


def valid_address(a: Address) -> bool:
    return a.street != "" and a.city != "" and a.state != "" and a.zip != ""


def unique_states(customers: list[Customer]) -> list[str]:
    seen = []
    for state in [c.address.state for c in customers]:
        if state not in seen:
            seen.append(state)
    return seen


def state_customer_count(customers: list[Customer], state: str) -> int:
    return len(customers_by_state(customers, state))


def points_for_purchase(amount: int, tier: str) -> int:
    base_points = amount // 100
    if tier == "gold":
        multiplier = 3
    elif tier == "silver":
        multiplier = 2
    elif tier == "bronze":
        multiplier = 1
    else:
        multiplier = 1
    return base_points * multiplier


def add_points(bal: LoyaltyBalance, amount: int, tier: str) -> LoyaltyBalance:
    earned = points_for_purchase(amount, tier)
    return LoyaltyBalance(bal.customer_id, bal.points + earned, bal.lifetime_points + earned)


def redeem_points(bal: LoyaltyBalance, points: int) -> LoyaltyBalance:
    available = bal.points
    to_redeem = min(points, available)
    return LoyaltyBalance(bal.customer_id, available - to_redeem, bal.lifetime_points)


def points_to_dollars(points: int) -> int:
    return points // 10


def find_loyalty_balance(balances: list[LoyaltyBalance], cust_id: int) -> Optional[LoyaltyBalance]:
    matches = [b for b in balances if b.customer_id == cust_id]
    return matches[0] if matches else None


def top_loyalty_customers(balances: list[LoyaltyBalance], n: int) -> list[LoyaltyBalance]:
    return list(sorted(balances, key=lambda b: b.points, reverse=True))[:n]


def customer_segment(records: list[PurchaseRecord], cust_id: int) -> str:
    spend = customer_total_spend(records, cust_id)
    cnt = customer_purchase_count(records, cust_id)
    if spend >= 100000 and cnt >= 10:
        return "vip"
    elif spend >= 50000 and cnt >= 5:
        return "regular"
    elif cnt >= 1:
        return "occasional"
    else:
        return "inactive"


def customers_by_segment(customers: list[Customer], records: list[PurchaseRecord], segment: str) -> list[Customer]:
    return [c for c in customers if customer_segment(records, c.id) == segment]


def segment_distribution(customers: list[Customer], records: list[PurchaseRecord]) -> dict:
    return {
        "vip": len(customers_by_segment(customers, records, "vip")),
        "regular": len(customers_by_segment(customers, records, "regular")),
        "occasional": len(customers_by_segment(customers, records, "occasional")),
        "inactive": len(customers_by_segment(customers, records, "inactive")),
    }


def years_since_last_purchase(records: list[PurchaseRecord], cust_id: int, current_year: int) -> int:
    purchases = customer_purchases(records, cust_id)
    if not purchases:
        return 999
    latest = functools.reduce(
        lambda best, r: r if r.year > best.year else best,
        purchases[1:],
        purchases[0],
    )
    return current_year - latest.year


def at_risk_customers(customers: list[Customer], records: list[PurchaseRecord], current_year: int, threshold_years: int) -> list[Customer]:
    return [c for c in customers if years_since_last_purchase(records, c.id, current_year) >= threshold_years]


def customer_retention_rate(customers: list[Customer], records: list[PurchaseRecord], current_year: int, threshold: int) -> int:
    total = len(customers)
    at_risk = len(at_risk_customers(customers, records, current_year, threshold))
    if total == 0:
        return 100
    return ((total - at_risk) * 100) // total


def update_customer_email(c: Customer, new_email: str) -> Customer:
    return Customer(c.id, c.name, new_email, c.tier, c.address, c.created_year)


def update_customer_address(c: Customer, new_addr: Address) -> Customer:
    return Customer(c.id, c.name, c.email, c.tier, new_addr, c.created_year)
